### DISCUSSION TITLE
Risk of multiplication overflow

### DISCUSSION THREAD

**Discussion by QGarchery** [morpho-org/morpho-blue-oracles#40] [General]
https://github.com/morpho-labs/morpho-blue-oracles/blob/0007653923d96e2d817f79b4a0e0fec87857e486/src/ChainlinkOracle.sol#L55
Can't this multiplication overflow and revert? We could use uniswap's 512-bit multiplication do circumvent this

_Originally posted by @Rubilmax in https://github.com/morpho-labs/morpho-blue-oracles/pull/1#discussion_r1343692737_

**Comment by QGarchery** [comment]
### Confirming the issue

Write `numerator = pB1*1e(fpB1) * pB2*1e(fpB2) * SCALE_FACTOR` as the multiplication that can overflow. We have `SCALE_FACTOR = 1e(36 + dQ1 + fpQ1 + fpQ2 - dB1 - fpB1 - fpB2)`, so:

`numerator = pB1 * pB2 * 1e(36 + dQ1 + fpQ1 + fpQ2 - dB1)`

If `fpQ1 = 18`, `fpQ2 = 18` and Q1 has 12 more decimals than B1, we get:
`numerator = pB1 * pB2 * 1e(84)`
So even if the prices `pB1` and `pB2` are 1 we get an overflow. This is not even taking into account the vault parameter.


### Possible solutions

I think that there are 3 possible ways to go further on this issue:
 1. **analyze the current oracles.** It may be ok to leave it like that if in practice it cannot overflow for all the first markets that will be deployed on Blue. Moreover, doing this analysis can help understand which quantities are contributing the most to the overflow.
 2. **use 512 bit multiplication and division.** This is the idea shown in #36, using a `mulDiv(a, b, c)` operation with full precision. This is not trivial because it requires to decide on how to split the whole multiplication in the numerator into the 2 factors `a` and `b`. 
 3. **use a sample amount.** The oracle would pass a sample amount of asset C, and we would use the feeds to determine how much it is worth in terms of asset B1 (call it `qB1`) and how much it is worth in terms of asset Q1 (call it `qQ1`) step by step. Each step would represent an amount of tokens, so it should not overflow. The final result can given by comparing `qB1` to `qQ1`. This solution would essentially require a rewrite of the oracle

**Comment by Rubilmax** [comment]
I don't get why in solution 3 we have to rewrite the oracle: can't we just assume sample = 1 and re-order multiplications and divisions?

But in this case the 512-bit multiplication would be closer to the real-number solution

**Comment by QGarchery** [comment]
> I don't get why in solution 3 we have to rewrite the oracle: can't we just assume sample = 1 and re-order multiplications and divisions?

If we assume we only compute amounts, `sample = 1` usually does not give enough precision. This is similar to [this comment](https://github.com/morpho-labs/morpho-blue-oracles/pull/19#discussion_r1358063876), where the amount you get after one step would be 1 whereas the price is 1.03

> But in this case the 512-bit multiplication would be closer to the real-number solution

True, but at the same time I'm not sure about the guarantees it gives. I guess it depends on how you split the factors `a` and `b`. The solution with the sample amounts should never overflow with a reasonable sample, because each step represents an amount

**Comment by MerlinEgalite** [comment]
I'd be in favor of option 2
